Enable OIDC based Authentication with apisix

ckhened opened the pull request:
## Description
The proposed changes enable OIDC (Open ID Connect) based user Authentication using APISIX API gateway and Keycloak Identity provider to OPEA apps.

This change introduces 2 helm charts:
1. helm-charts/auth-apisix/apisix-helm: This helm chart installs APISIX api gateway and APISIX ingress controller
2. helm-charts/auth-apisix/apis-crd-helm: This helm chart contains CRDs for authenticated APIs that needs to be published in API gateway

The Readme file in helm-charts/auth-apisix/README.md gives instructions to install keycloak, apisix and API CRDs

APISIX is apache licensed open source API gateway which is light weight, delivers high performance. It can work with docker or kubernetes and with any service mesh within kubernetes.


## Issues
n/a


## Type of change

List the type of change like below. Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X ] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would break existing design and interface)

## Dependencies
1. apisix api gateway
2. apisix-ingress-controller
3. keycloak


## Tests
Verified locally on my test system

hshen14 commented on helm-charts/auth-apisix/README.md:
````diff
@@ -0,0 +1,56 @@
+Steps to start keycloak from official keycloak helm chart
+
+```sh
+# Prerequisite: Create a PersistentVolume of 9Gi for keycloak-postgress with RWO access (to persist updated keycloak configuration)
````
Do we need to add a manifest to show how to create such persistentvolume or refer to some link? It's better that the readme is self-contained.

lianhao commented on helm-charts/auth-apisix/README.md:
```diff
@@ -0,0 +1,77 @@
+helm dependency update
+
+# Install apisix
+helm install auth-apisix . --create-namespace --namespace auth-apisix
```
The apisix-helm seems doesn't contain any K8S resources. It just install a third party apisix helm chart with our config values, right? If so, we would suggest to put the values files here, and remove the whole apisix-helm directory. So there is no need to worry about helm chart version/dependency in the future.

ckhened replied:
incorporated this and made the change in directory structure and readme

lianhao commented on helm-charts/auth-apisix/apis-crd-helm/values.yaml:
```diff
@@ -0,0 +1,20 @@
+# APISIX chatqna api config
+chatqna:
+  namespace: <namespace in which your chatqna service is running>
+  hostname: <hostname for external access>
```
what does this hostname for external access mean?

ckhened replied:
added more details for clarity

lianhao commented on helm-charts/auth-apisix/README.md:
````diff
@@ -0,0 +1,77 @@
+```sh
+# Uninstall apisix
+helm uninstall auth-apisix-crds --namespace auth-apisix
+helm uninstall auth-apisix --namespace auth-apisix
````
I thing CRD installed by apisix-helm won't be deleted by `helm uninstall`. So we need to add a new notes for users of how to clean up those CRDs.

ckhened replied:
Added note to cleanup the apisix crds

lianhao commented on helm-charts/auth-apisix/README.md:
```diff
@@ -0,0 +1,87 @@
+While accessing the published APIs, the HTTP Authorization header of the request should contain the Access token provided by Identity provider as 'Bearer \<Access Token\>' </br></br>
+The access token, refresh token, userinfo and user roles can be obtained by invoking OIDC auth endpoint through UI or token endpoint through curl and providing user credentials. </br>
+
+## Uninstall
```
Please also add a paragraph of how to use the APISIX gateway to cnnect to chatqna after following the above steps.

mkbhanda replied:
@ckhened I would like this PR organized a little differently. 
1) APISIX is a gateway/proxy, certainly a infrastructure related. Would you kindly make a directory under GenAIInfra called Gateway and then under that add APISIX.
so GenAIInfra/Gateway/APISIX/helmchart ...

2) Any new/additional Authn&Z stuff lets put under https://github.com/opea-project/GenAIInfra/tree/main/authN-authZ. We already have Keycloak there is my understanding.

3) Under APISIX add docs on how to include authN&Z.

Please work with @Ruoyu-y to discuss restructuring. 

Thank you for addressing the other comments.

ckhened replied:
@mkbhanda Sure. I'll sync up with @Ruoyu-y for restructuring

lianhao replied:
How about we create 2 subdirectories under the top level authN-authZ dir, one for the current istio gateway implemenation, and one for the apisix implemenation. And also add a top level README under authN-authZ for users to reference different options? @Ruoyu-y any comment?

Ruoyu-y replied:
@ckhened authN-authZ is originally defined for all authentication and authorization related stuff. If you are leveraging the same keycloak, we can have a directory for OIDC provider that can be used by both implementations. I suggest to make the directory structure like this:
```
authN-authZ
├── auth-istio
│   ├── ...
│   ├── docs/
│   └── README.md // Current readme file for Istio based implementation
├── auth-apisix
│   ├── ...
│   └── README.md // readme file for APISIX based implementation
├── auth-oidc-provider
│   └── keycloak_install.yaml // Or even use the community installation method and note that in both readmes: 
│                                              // kubectl create -f https://raw.githubusercontent.com/keycloak/keycloak- 
│                                              // quickstarts/latest/kubernetes/keycloak.yaml
└── README.md // Overall readme file for authentication and authorization and links to different implementations
```
@lianhao @mkbhanda Kindly share your thoughts here.

lianhao replied:
@Ruoyu-y how about we keep the keycloak installation separate in each different scenario, because I see the installation options of keycloak for istio-gateway and apisix is a little bit different. Also we can't make sure that in the future a 3rd auth scenario would use the same keycloak as well.

Ruoyu-y replied:
> @Ruoyu-y how about we keep the keycloak installation separate in each different scenario, because I see the installation options of keycloak for istio-gateway and apisix is a little bit different. Also we can't make sure that in the future a 3rd auth scenario would use the same keycloak as well.

Sure. Let's put keycloak installation and guide in each directory.

Ruoyu-y replied:
BTW, i just created a draft PR #353 to restructure the directory. @ckhened please check if possible to add apisix implementation under such structure.

ckhened replied:
@Ruoyu-y with the directory structure in your PR, apisix can be added without any problem. I'll wait for your PR to be merged and then add my changes in the authN-authZ directory. Meanwhile i'll fix all the review comments

ckhened replied:
Moved auth-apisix to new directory structure

lianhao reviewed (commented):
Since this could NOT be covered by CI test due to the pre-requests, maybe we could move this to the top level folder [authN-authZ](https://github.com/opea-project/GenAIInfra/tree/main/authN-authZ), because this also resolves authentication & authorization.

ckhened commented on helm-charts/auth-apisix/README.md:
````diff
@@ -0,0 +1,85 @@
+
+```sh
+# Install apisix api gateway and ingress controller
+helm install auth-apisix apisix/apisix -f values_apisix_gw.yaml --create-namespace --namespace auth-apisix
````
Please suggest if any alternate approach should be taken (like adding a copy of the remote helm chart ?)

lianhao replied:
It could be something like the following
```
helm repo add apisix https://charts.apiseven.com
helm repo update
helm install apisix apisix/apisix --create-namespace  --namespace .. --version ... -f ...
```

ckhened replied:
added the details

ckhened commented on helm-charts/auth-apisix/README.md:
````diff
@@ -0,0 +1,85 @@
+```
+
+</br></br>
+Apisix helm chart provides configs to change the service type to other options like LoadBalancer (apisix.service.type) and externalTrafficPolicy to 'local'(apisix.service.externalTrafficPolicy). These can be added in apisix-helm/values.yaml </br></br>
````
fixed this

ckhened commented on helm-charts/auth-apisix/values_apisix_gw.yaml:
```diff
@@ -0,0 +1,21 @@
+# SPDX-License-Identifier: Apache-2.0
+
+# APISIX Namespace
+namespace: auth-apisix
```
that is not needed. removed it

ckhened commented on helm-charts/auth-apisix/README.md:
```diff
@@ -0,0 +1,85 @@
+</br></br>
+Apisix helm chart provides configs to change the service type to other options like LoadBalancer (apisix.service.type) and externalTrafficPolicy to 'local'(apisix.service.externalTrafficPolicy). These can be added in apisix-helm/values.yaml </br></br>
+While accessing the published APIs, the HTTP Authorization header of the request should contain the Access token provided by Identity provider as 'Bearer \<Access Token\>' </br></br>
+The access token, refresh token, userinfo and user roles can be obtained by invoking OIDC auth endpoint through UI or token endpoint through curl and providing user credentials. </br>
```
added example

lianhao commented:
Also please fix the DCO error, thx

Ruoyu-y commented on helm-charts/auth-apisix/README.md:
```diff
@@ -0,0 +1,113 @@
+The access token, refresh token, userinfo and user roles can be obtained by invoking OIDC auth endpoint through UI or token endpoint through curl and providing user credentials. </br></br>
+
+Below steps can be followed to get access token from keycloak and access the APISIX published ChatQnA API through curl
+
```
Please help explain your authentication and authorization scenarios here.

ftian1 replied:
I agree the proposal of the above folder layout.

lianhao commented:
@ckhened PR #353 is merged. Please refactor based on that. Thanks!

lianhao reviewed (changes requested):
Please remove duplicated files under helm-charts directory. Also please add reference in file authN-authZ/README.md

ckhened commented:
Moved auth-apisix dir to authN-authZ, fixed CI checks and incorporated other comments.
I'll add authentication and authorization scenarios examples (comment from @Ruoyu-y ) in the next commit

lianhao commented on authN-authZ/auth-apisix/README.md:
```diff
@@ -0,0 +1,115 @@
+
+</br>
+Apisix helm chart provides configs to change the service type to other options like LoadBalancer (apisix.service.type) and externalTrafficPolicy to 'local'(apisix.service.externalTrafficPolicy). These can be added in values_apisix_gw.yaml </br></br>
+While accessing the published APIs, the HTTP Authorization header of the request should contain the Access token provided by Identity provider as 'Bearer \<Access Token\>' </br></br>
```
`'Bearer \<Access Token\>'` is rendered by github webpage as `'Bearer \'`. Maybe we should use use the following instead?
```suggestion
... ... ... `Bearer <Access Token\>`
```

ckhened replied:
corrected!

ckhened commented on authN-authZ/auth-apisix/README.md:
```diff
@@ -0,0 +1,115 @@
+# Authentication and Authorization with APISIX and OIDC based Identity provider (Keycloak)
```
updated!

Ruoyu-y commented:
> Moved auth-apisix dir to authN-authZ, fixed CI checks and incorporated other comments. I'll add authentication and authorization scenarios examples (comment from @Ruoyu-y ) in the next commit

This will not include in this PR, correct?
